ktor-shared/ktor-resources/common/src/io/ktor/resources/serialisation/ResourcesFormat.kt

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2014-2021 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package io.ktor.resources.serialisation
+
+import io.ktor.http.*
+import io.ktor.resources.*
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.modules.*
+
+/**
+ * A format to (de)serialize resources instances
+ */
+@OptIn(ExperimentalSerializationApi::class)
+public class ResourcesFormat(
+    override val serializersModule: SerializersModule = EmptySerializersModule
+) : SerialFormat {
+
+    /**
+     * A query parameter description
+     */
+    public data class Parameter(
+        val name: String,
+        val isOptional: Boolean
+    )
+
+    /**
+     * Builds a path pattern for a given [serializer]
+     */
+    public fun <T> encodeToPathPattern(serializer: KSerializer<T>): String {
+        val pathBuilder = StringBuilder()
+
+        var current: SerialDescriptor? = serializer.descriptor
+        while (current != null) {
+            val path = current.annotations.filterIsInstance<Resource>().first().path
+            val addSlash = pathBuilder.isNotEmpty() && !pathBuilder.startsWith('/') && !path.endsWith('/')
+            if (addSlash) {
+                pathBuilder.insert(0, '/')
+            }
+            pathBuilder.insert(0, path)
+
+            val membersWithAnnotations = current.elementDescriptors.filter { it.annotations.any { it is Resource } }
+            if (membersWithAnnotations.size > 1) {
+                throw ResourceSerializationException("There are multiple parents for resource ${current.serialName}")
+            }
+            current = membersWithAnnotations.firstOrNull()
+        }
+
+        if (pathBuilder.startsWith('/')) {
+            pathBuilder.deleteAt(0)
```

Comment:
I was playing around with new locations and ran into unexpected behaviour caused by this.
What was the reasoning for doing this? I don't mind it but I was quite puzzled when my tests were passing when they shouldn't have.